flang/docs/ParallelFortranRuntime.md

```diff
@@ -0,0 +1,1987 @@
+  * 1d assumed-shape array of type `integer(c_intmax_t)`
+  * correspond to the cosubscripts appearing in a _coindexed-named-object_ reference
+* **`value`** or `current_image_buffer`
+  * Argument for `put` and `get` operations
+  * assumed-rank array of `type(*)` or `type(c_ptr)`
+  * It is the value to be sent in a `put` operation, and is assigned the value
+    retrieved in the case of a `get` operation
+* **`image_num`**
+  * an argument identifying the image to be communicated with
+  * is permitted to identify the current image
+  * the image index is always relative to the initial team, unless otherwise specified
+
+### `prif_allocate_coarray`
+
+**Description**: This procedure allocates memory for a coarray. 
+  This call is collective over the current team.  Calls to
+  `prif_allocate_coarray` will be inserted by the compiler when there is an explicit
+  coarray allocation or at the beginning of a program to allocate space for
+  statically declared coarrays in the source code. The PRIF implementation will
+  store the coshape information in order to internally track it during the
+  lifetime of the coarray.
+    
+```
+subroutine prif_allocate_coarray( &
+    lcobounds, ucobounds, lbounds, ubounds, element_length, &
+    final_func, coarray_handle, allocated_memory, &
+    stat, errmsg, errmsg_alloc)
+  integer(kind=c_intmax_t), intent(in) :: lcobounds(:), ucobounds(:)
+  integer(kind=c_intmax_t), intent(in) :: lbounds(:), ubounds(:)
+  integer(kind=c_size_t), intent(in) :: element_length
+  type(c_funptr), intent(in) :: final_func
+  type(prif_coarray_handle), intent(out) :: coarray_handle
+  type(c_ptr), intent(out) :: allocated_memory
+  integer(c_int), intent(out), optional :: stat
+  character(len=*), intent(inout), optional :: errmsg
+  character(len=:), intent(inout), allocatable, optional :: errmsg_alloc
+end subroutine
+```
+**Further argument descriptions**:
+
+* **`lcobounds` and `ucobounds`**: Shall be the lower and upper bounds of the
+  codimensions of the coarray being allocated. Shall be 1d arrays with the
+  same dimensions as each other. The cobounds shall be sufficient to have a
+  unique index for every image in the current team.
+  I.e. `product(ucobounds - lcobounds + 1) >= num_images()`.
+* **`lbounds` and `ubounds`**: Shall be the the lower and upper bounds of the
+  current image's portion of the array. Shall be 1d arrays with the same dimensions as
+  each other.
+* **`element_length`**: size of a single element of the array in bytes
+* **`final_func`**: Shall be the C address of a procedure that is interoperable, or
+  `C_NULL_FUNPTR`. If not null, this procedure will be invoked by the PRIF implementation
+  once by each image at deallocation of this coarray, before the storage is released.
+  The procedure's interface shall be equivalent to the following Fortran interface
+  ```
+  subroutine coarray_cleanup(handle, stat, errmsg) bind(C)
+    type(prif_coarray_handle), pointer, intent(in) :: handle
```

Comment:
Quoted from @jeffhammond's response:

> `final_func` takes the coarray handle as `intent(in)`. Is there no use case for `final_func` modifying the state inside of the coarray handle?

Reply:
**LBL response:** `prif_coarray_handle` represents a reference to a coarray descriptor, not the descriptor itself. The state of the coarray descriptor that the handle refers to may be changed by the finalizer, but not the coarray handle itself.